# Set variables and set up environment
```{bash, engine.opts='-l', eval=FALSE}
POP="EUR"
PHENO="SPC"
GENO="HRC"
MODEL="1Mbase"
PROJECT="${POP}_${PHENO}_${GENO}_${MODEL}"
THREADS=2
mkdir -p data/$PROJECT/output/fin_peds
mkdir -p data/$PROJECT/output/graphs
mkdir -p data/$PROJECT/output/grm_chrs
mkdir -p data/$PROJECT/output/grm_ranges
mkdir -p data/$PROJECT/output/grms
mkdir -p data/$PROJECT/output/hsqs
mkdir -p data/$PROJECT/output/mgrms
mkdir -p data/$PROJECT/output/results
mkdir -p data/$PROJECT/input/

pushd data/$PROJECT/input/
ln -s ../../../fin_genotype/${POP}_${GENO}/* ./
ln -s ../../../fin_RNA_seq/${POP}_${PHENO}/* ./
popd

pushd data/$PROJECT/output/grms
ln -s ../../../../grms/${POP}_${GENO}/* ./
popd
```

```{r, eval=FALSE}
PROJECT="AFR_SPC_HRC_1Mbase"
```

# Prepare and subset VCF and BED files
Use the code from trans-eQTL analysis, which subset the data into European frontal 
cortex samples and regressed out covariates

# Remove nonpolymorphic variants
```{bash, engine.opts='-l', eval=FALSE}
bcftools view -e 'COUNT(GT="AA")=N_SAMPLES || COUNT(GT="RR")=N_SAMPLES' \
    data/$PROJECT/input/nonpoly_subset.vcf.gz -o data/$PROJECT/input/subset.vcf && \
gzip data/$PROJECT/input/subset.vcf
```
# Convert VCF to PED
```{bash, engine.opts='-l', eval=FALSE}
# Set family ID all to 0
plink --vcf data/${PROJECT}/input/subset.vcf.gz --recode --out data/${PROJECT}/input/ped_file --const-fid --make-bed
```

# Generate complete GRM
We split the GRM by chromosome, and then combine it to generate a complete GRM.
```{bash, engine.opts='-l', eval=FALSE}
> data/${PROJECT}/output/grm_chrs/all_chrs.txt
for ((i=1;i<=22;i++)); do
    gcta64 --make-grm-bin --bfile data/${PROJECT}/input/ped_file --make-grm-alg 0 --chr $i --out data/${PROJECT}/output/grms/ped_file$i --thread-num ${THREADS} && \
    printf "data/%s/output/grms/ped_file%s\n" $PROJECT $i >> data/${PROJECT}/output/grm_chrs/all_chrs.txt
done

gcta64 --mgrm data/${PROJECT}/output/grm_chrs/all_chrs.txt --make-grm-bin --out data/${PROJECT}/output/grms/complete
```

# Generate Big K GRM
We modify the complete GRM by setting unrelated pairs to 0, < 0.025 as the threshold.  
```{bash, engine.opts='-l', eval=FALSE}
gcta64 --grm data/${PROJECT}/output/grms/complete --thread-num $THREADS --make-bK 0.025 \
    --out data/${PROJECT}/output/grms/bigk
```

# Generate histogram and report how many pairs have a relatedness > 0.025, > 0.05.
```{r, eval=FALSE}
# R script to read the GRM binary file
ReadGRMBin=function(prefix, AllN=F, size=4){
    sum_i=function(i){
        return(sum(1:i))
    }
    BinFileName=paste(prefix,".grm.bin",sep="")
    NFileName=paste(prefix,".grm.N.bin",sep="")
    IDFileName=paste(prefix,".grm.id",sep="")
    id = read.table(IDFileName)
    n=dim(id)[1]
    BinFile=file(BinFileName, "rb");
    grm=readBin(BinFile, n=n*(n+1)/2, what=numeric(0), size=size)
    NFile=file(NFileName, "rb");
    if(AllN==T){
        N=readBin(NFile, n=n*(n+1)/2, what=numeric(0), size=size)
    }
    else N=readBin(NFile, n=1, what=numeric(0), size=size)
    i=sapply(1:n, sum_i)
    return(list(diag=grm[i], off=grm[-i], id=id, N=N))
}

grm = ReadGRMBin(paste('data/', PROJECT, '/output/grms/complete', sep=''))

library('ggplot2')
png(paste('data/', PROJECT, '/output/graphs/grm_histogram.png', sep=''))
ggplot(data = data.frame(off=grm[['off']])) + geom_histogram(mapping = aes(x=off), bins=100)
dev.off()

png(paste('data/', PROJECT, '/output/graphs/trunc_grm_histogram.png', sep=''))
ggplot(data = data.frame(off=grm[['off']][grm[['off']] < 0.05])) + geom_histogram(mapping = aes(x=off), bins=100)
dev.off()
# 593 pairs with GR > 0.025
# 5 pairs with GR > 0.05
```

# Subset GRM to filter out cryptic relatedness (GR > 0.05)
```{bash, engine.opts='-l', eval=FALSE}
gcta64 --grm data/${PROJECT}/output/grms/complete --grm-cutoff 0.05 --make-grm-bin --out data/${PROJECT}/output/grms/unrelated
# running diff on the complete.grm.id and unrelated.grm.id files gives us 
# the 5 ommitted samples

# CMC_MSSM_307
# CMC_PITT_077
# Br1977
# Br1876
# Br2173

# We write them to a file to use in GREML
printf '0 CMC_MSSM_307\n0 CMC_PITT_077\n0 Br1977\n0 Br1876\n0 Br2173' > data/${PROJECT}/input/removed_samples.txt
```

# Run GREML
First, we need to generate a phenotype file.
```{r, eval=FALSE}
# 1 Mbase window
window_size = 1e6
bed_file = read.table(pipe(paste('zcat data/', PROJECT, '/input/sPsychENCODE-gene.BED.gz | sed -e 1s/.//', sep='')), 
                      header = TRUE, sep='\t')
bed_file[, 'Chr'] = sapply(X = bed_file[, 'Chr'], FUN = function(chr_str) {
                           return(gsub('[^0-9]', '', chr_str))
                      })
bed_file[bed_file[, 'Chr'] == '', 'Chr'] = '0'
fam_id = rep(x = 0, times=dim(bed_file)[2] - 6)
sample_id = scan(text = readLines(paste('data/', PROJECT, '/input/sPsychENCODE-gene.BED.gz', sep=''), 1), what = "", 
                 quiet = TRUE)[-(1:6)]
phenotype_id = bed_file[,'pid']
phenotypes = bed_file[,7:dim(bed_file)[2]]
phenotypes = t(as.matrix(phenotypes))
colnames(phenotypes) = phenotype_id
rownames(phenotypes) = sample_id

window_start = bed_file[, 'start'] - window_size
window_start[window_start < 0] = 0

window_end = bed_file[, 'end'] + window_size

phenotype_info = cbind(phenotype_id, bed_file[,c('Chr', 'start', 'end')], window_start, window_end)

write.table(phenotype_info, paste('data/', PROJECT, '/input/phenotype_ids', sep=''), col.names=FALSE, row.names=FALSE, quote=FALSE)

pheno_file = cbind(fam_id, sample_id, phenotypes)
write.table(pheno_file, paste('data/', PROJECT, '/input/phenotype', sep=''), sep='\t', col.names=FALSE, row.names=FALSE, 
            quote=FALSE)
```

# Automating for all SNPs

## Necessary steps
1. Generate subsets of SNPs in cis and trans regions from PLINK  
2. Generate GRM of the cis and trains subsets  
3. Generate mgrm.txt
4. Run GREML  
Script is written in run\_all.sh

# Combine the HSQ files and return statistics
Scripts for compiling data from HSQ files are in scripts/hsq_\*

# Notes
[The Genetic Architecture of Gene Expression in Peripheral Blood](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5294670/)  
The above paper treats cis-eQTLs as those where the SNP and effect are on the same 
chromosome, and trans if not:  
"cis-eQTL were defined to be those associations where the SNP was located on the same 
chromosome as the gene, and trans-eQTL the complement of this."

EUR: 5 pairs > 0.05
1Mbase window
cis 0.3375173
trans 0.6624827
avg snp h2 0.1085098
5345/24905 significant heritable SNPs

chromosome window
cis 0.3980837
trans 0.6019163
avg snp h2 0.1285103
2028/24905 significant heritable SNPs

whole genome
avg snp h2 0.06809028
770/24905 significant heritable SNPs

from bKsK
avg snp h2 0.06298553
narrow sense h2 0.1926015
434/24905 significant heritable SNPs

AFR: 75 pairs > 0.05
322 pairs > 0.025
0.05 pairs statistics:
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
0.05006 0.05679 0.06271 0.10121 0.07700 0.94868
1Mbase window
cis 0.2001722
trans 0.7998278
avg snp h2 0.3041369
3717/24905 significant heritable SNPs

chromosome window
cis 0.3348791
trans 0.6651209
avg snp h2 0.3427922
1495/24905 significant heritable SNPs

whole genome
avg snp h2 0.2510293
1770/24905 significant heritable SNPs

from bKsK
avg snp h2 0.1160139
narrow sense h2 0.403915
345/24905 significant heritable SNPs

[The Genetic Architecture of Gene Expression in Peripheral Blood](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5294670/)  
That Lloyd-Jones paper  

[Using Extended Genealogy to Estimate Components of Heritability for 23 Quantitative and Dichotomous Traits](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003520)  
Paper on estimating narrow-sense heritability

[Narrow-sense heritability estimation of complex traits using identity-by-descent information](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6221881/)  
A friendly explanation of narrow-sense heritability estimation  

[Neale Lab Blog Post on heritability](http://www.nealelab.is/blog/2017/9/13/heritability-101-what-is-heritability)  
Helpful blog post on the basics of heritability, with links to their more indepth 
posts about how their lab estimated heritability  

[Genome-wide eQTLs and heritability for gene expression traits in unrelated individuals](https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-15-13)
Another paper on estimating cis and trans contribution to heritability

[A characterization of cis- and trans-heritability of RNA-Seq-based gene expression](https://www.nature.com/articles/s41431-019-0511-5#Sec20)  
Another paper on estimating cis and trans contribution to heritability

[Heritability and Tissue Specificity of Expression Quantitative Trait Loci](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.0020172)  
Tangentially related paper, examining heritability in the context of cis and trans eQTLs  

[The influence of genetic variation on gene expression](https://genome.cshlp.org/content/17/12/1707.full.html)  
An overview paper about the topic  

## Todo
Find index cis-eQTLs and find optimal HCP factors
Split the HSQ processing scripts into their own files

Unrelated  
V(nxn) = GRM_cis sigma^2_cis + GRM_trans sigma^2_trans + I(nxn) sigma^2_epsilon  
V(nxn) = GRM sigma^2 + I(nxn) sigma^2_epsilon  

Related  
V(nxn) = GRM(>0.025) sigma^2_1 + GRM sigma^2_2 + I(nxn) sigma^2_epsilon  
Try to partition  

