# Set variables and set up environment
```{bash, engine.opts='-l', eval=FALSE}
PROJECT="1MBase"
THREADS=2
mkdir -p data/$PROJECT/output/fin_peds
mkdir -p data/$PROJECT/output/graphs
mkdir -p data/$PROJECT/output/grm_chrs
mkdir -p data/$PROJECT/output/grm_ranges
mkdir -p data/$PROJECT/output/grms
mkdir -p data/$PROJECT/output/hsqs
mkdir -p data/$PROJECT/output/mgrms
mkdir -p data/$PROJECT/output/results
mkdir -p data/$PROJECT/input/
```

```{r, eval=FALSE}
PROJECT="1MBase"
```

# Prepare and subset VCF and BED files
Use the code from trans-eQTL analysis, which subset the data into European frontal 
cortex samples and regressed out covariates

# Convert VCF to PED
```{bash, engine.opts='-l', eval=FALSE}
# Set family ID all to 0
plink --vcf data/${PROJECT}/input/subset.vcf.gz --recode --out data/${PROJECT}/input/ped_file --const-fid --make-bed
```

# Generate complete GRM
We split the GRM by chromosome, and then combine it to generate a complete GRM.
```{bash, engine.opts='-l', eval=FALSE}
> data/${PROJECT}/output/grm_chrs/all_chrs.txt
for ((i=1;i<=22;i++)); do
    gcta64 --make-grm-bin --bfile data/${PROJECT}/input/ped_file --make-grm-alg 0 --chr $i --out data/${PROJECT}/output/grms/ped_file$i --thread-num ${THREADS}
    printf "data/%s/output/grms/ped_file%s" $PROJECT $i >> data/${PROJECT}/output/grm_chrs/all_chrs.txt
done

gcta64 --mgrm data/${PROJECT}/output/grm_chrs/all_chrs.txt --make-grm-bin --out data/${PROJECT}/output/grms/complete
```

# Generate histogram and report how many pairs have a relatedness > 0.025, > 0.05.
```{r, eval=FALSE}
# R script to read the GRM binary file
ReadGRMBin=function(prefix, AllN=F, size=4){
    sum_i=function(i){
        return(sum(1:i))
    }
    BinFileName=paste(prefix,".grm.bin",sep="")
    NFileName=paste(prefix,".grm.N.bin",sep="")
    IDFileName=paste(prefix,".grm.id",sep="")
    id = read.table(IDFileName)
    n=dim(id)[1]
    BinFile=file(BinFileName, "rb");
    grm=readBin(BinFile, n=n*(n+1)/2, what=numeric(0), size=size)
    NFile=file(NFileName, "rb");
    if(AllN==T){
        N=readBin(NFile, n=n*(n+1)/2, what=numeric(0), size=size)
    }
    else N=readBin(NFile, n=1, what=numeric(0), size=size)
    i=sapply(1:n, sum_i)
    return(list(diag=grm[i], off=grm[-i], id=id, N=N))
}

grm = ReadGRMBin(paste('data/', PROJECT, '/output/grms/complete', sep=''))

library('ggplot2')
png(paste('data/', PROJECT, '/output/graphs/grm_histogram.png', sep=''))
ggplot(data = data.frame(off=grm[['off']])) + geom_histogram(mapping = aes(x=off), bins=100)
dev.off()

png(paste('data/', PROJECT, '/output/graphs/trunc_grm_histogram.png', sep=''))
ggplot(data = data.frame(off=grm[['off']][grm[['off']] < 0.05])) + geom_histogram(mapping = aes(x=off), bins=100)
dev.off()
# 593 pairs with GR > 0.025
# 5 pairs with GR > 0.05
```

# Subset GRM to filter out cryptic relatedness (GR > 0.05)
```{bash, engine.opts='-l', eval=FALSE}
gcta64 --grm data/${PROJECT}/output/grms/complete --grm-cutoff 0.05 --make-grm-bin --out data/${PROJECT}/output/grms/unrelated
# running diff on the complete.grm.id and unrelated.grm.id files gives us 
# the 5 ommitted samples

# CMC_MSSM_307
# CMC_PITT_077
# Br1977
# Br1876
# Br2173

# We write them to a file to use in GREML
printf '0 CMC_MSSM_307\n0 CMC_PITT_077\n0 Br1977\n0 Br1876\n0 Br2173' > data/${PROJECT}/input/removed_samples.txt
```

# Run GREML
First, we need to generate a phenotype file.
```{r, eval=FALSE}
# 1 Mbase window
window_size = 1e6
bed_file = read.table(pipe(paste('zcat data/', PROJECT, '/input/sPsychENCODE-gene.BED.gz | sed -e 1s/.//', sep='')), 
                      header = TRUE, sep='\t')
bed_file[, 'Chr'] = sapply(X = bed_file[, 'Chr'], FUN = function(chr_str) {
                           return(gsub('[^0-9]', '', chr_str))
                      })
bed_file[bed_file[, 'Chr'] == '', 'Chr'] = '0'
fam_id = rep(x = 0, times=dim(bed_file)[2] - 6)
sample_id = scan(text = readLines(paste('data/', PROJECT, '/input/sPsychENCODE-gene.BED.gz', sep=''), 1), what = "", 
                 quiet = TRUE)[-(1:6)]
phenotype_id = bed_file[,'pid']
phenotypes = bed_file[,7:dim(bed_file)[2]]
phenotypes = t(as.matrix(phenotypes))
colnames(phenotypes) = phenotype_id
rownames(phenotypes) = sample_id

window_start = bed_file[, 'start'] - window_size
window_start[window_start < 0] = 0

window_end = bed_file[, 'end'] + window_size

phenotype_info = cbind(phenotype_id, bed_file[,c('Chr', 'start', 'end')], window_start, window_end)

write.table(phenotype_info, paste('data/', PROJECT, '/input/phenotype_ids', sep=''), col.names=FALSE, row.names=FALSE, quote=FALSE)

pheno_file = cbind(fam_id, sample_id, phenotypes)
write.table(pheno_file, paste('data/', PROJECT, '/input/phenotype', sep=''), sep='\t', col.names=FALSE, row.names=FALSE, 
            quote=FALSE)
```

# Automating for all SNPs

## Necessary steps
1. Generate subsets of SNPs in cis and trans regions from PLINK  
2. Generate GRM of the cis and trains subsets  
3. Generate mgrm.txt
4. Run GREML  
Script is written in run\_all.sh

# Combine the HSQ files and return statistics
```{r, eval=FALSE}
temp = unlist(list.files(path=paste("data/", PROJECT, "/output/hsqs", sep=""), pattern="*.hsq"))
hsqs = t(sapply(X=temp, function(file_name) {
                  hsq = read.table(paste('hsqs/', file_name, sep=''), header=TRUE, nrows=4)
                  # Gene ID, Cis, Trans, Cis+Trans, Residual, Phenotype
                  line = c(substring(file_name, 1, nchar(file_name) - 4), 
                           hsq[1,2], hsq[2,2], hsq[1,2] + hsq[2,2], hsq[3,2], hsq[4,2])
                  return(line)
            }))
colnames(hsqs) = c("ID", "Cis", "Trans", "SNP", "Res", "Phe")
rownames(hsqs) = hsqs[,"ID"]
hsqs = as.data.frame(hsqs)
hsqs = transform(hsqs, Cis=as.numeric(Cis), Trans=as.numeric(Trans), 
                 SNP=as.numeric(SNP), Res=as.numeric(Res), Phe=as.numeric(Phe))
write.table(hsqs, paste("data/", PROJECT, "/output/all_variance.txt", sep=""), quote=FALSE, row.names=FALSE, col.names=TRUE)

cis_prop = sum(hsqs[, "Cis"])/sum(hsqs[, "SNP"]) # 0.3395873
trans_prop = sum(hsqs[, "Trans"])/sum(hsqs[, "SNP"]) # 0.6604127
avg_snp_h2 = mean(apply(X=hsqs, MARGIN=1, function(row) as.numeric(row[4])/as.numeric(row[6]))) # 0.1087606
```

# Notes
[The Genetic Architecture of Gene Expression in Peripheral Blood](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5294670/)  
The above paper treats cis-eQTLs as those where the SNP and effect are on the same 
chromosome, and trans if not:  
"cis-eQTL were defined to be those associations where the SNP was located on the same 
chromosome as the gene, and trans-eQTL the complement of this."

